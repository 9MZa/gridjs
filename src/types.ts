import { ComponentChild } from 'preact';
import Row from './row';
import { SortConfig } from './view/plugin/sort/sort';

export type ProtoExtends<T, U> = U & Omit<T, keyof U>;

export type OneDArray<T> = T[];
export type TwoDArray<T> = T[][];

// Table cell types
export type TCell = number | string | boolean | ComponentChild | HTMLElement;
export type TData = TwoDArray<TCell> | OneDArray<{ [key: string]: TCell }>;

// Table header cell type
export interface TColumn {
  id?: string;
  name: string;
  width?: string;
  sort?: SortConfig;
  children?: OneDArray<TColumn>;
  formatter?: (cell: TCell, row: Row, column: TColumn) => ComponentChild;
}

// Comparator function for the sorting plugin
export type Comparator<T> = (a: T, b: T) => number;

export interface TColumnSort {
  index: number;
  // 1 ascending, -1 descending
  direction?: 1 | -1;
}

// container status
export enum Status {
  Init,
  Loading,
  Loaded,
  Rendered,
  Error,
}

export type CSSDeclaration = {
  [key: string]: string;
};
