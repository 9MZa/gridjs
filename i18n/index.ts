import esES from './es_ES';
import frFR from './fr_FR';
import itIT from './it_IT';

export {
  esES,
  frFR,
  itIT
};
