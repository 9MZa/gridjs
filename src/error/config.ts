import BaseError from './base';

class ConfigError extends BaseError {}
export default ConfigError;
