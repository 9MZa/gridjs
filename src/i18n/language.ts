import enUS from './en_US';
export type Language = { [key: string]: string | Language };

export class Translator {
  private readonly _language: Language;
  private readonly _defaultLanguage: Language;

  constructor(language?: Language) {
    this._language = language;
    this._defaultLanguage = enUS;
  }

  /**
   * Tries to split the message with "." and find
   * the key in the given language
   *
   * @param message
   * @param lang
   */
  getString(message: string, lang: Language): string {
    if (!lang || !message) return null;

    const splitted = message.split('.');
    const key = splitted[0];

    if (lang[key]) {
      const val = lang[key];

      if (typeof val === 'string') {
        return val;
      } else {
        return this.getString(splitted.slice(1).join('.'), val);
      }
    }

    return null;
  }

  translate(message: string): string {
    const translated = this.getString(message, this._language);

    if (translated) {
      return String(translated);
    }

    return this.getString(message, this._defaultLanguage);
  }
}

export function useTranslator(translator: Translator) {
  return function (message: string): string {
    return translator.translate(message);
  };
}
