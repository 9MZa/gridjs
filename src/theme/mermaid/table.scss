table.gridjs {
  &-table {
    border-collapse: collapse;
    text-align: left;
    display: table;
    margin: 0;
    padding: 0;
    overflow: auto;
  }
}
