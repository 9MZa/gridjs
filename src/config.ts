import { OneDArray, ProtoExtends, TCell, TColumn, TwoDArray } from './types';
import Storage from './storage/storage';
import Pipeline from './pipeline/pipeline';
import Tabular from './tabular';
import { SearchConfig } from './view/plugin/search/search';
import { PaginationConfig } from './view/plugin/pagination';
import Header from './header';
import { ServerStorageOptions } from './storage/server';
import Dispatcher from './util/dispatcher';
import { GenericSortConfig } from './view/plugin/sort/sort';
import { Language, Translator } from './i18n/language';
import { createRef, RefObject } from 'preact';
import StorageUtils from './storage/storageUtils';
import PipelineUtils from './pipeline/pipelineUtils';

// Config type used internally
export interface Config {
  dispatcher?: Dispatcher<any>;
  /** container element that is used to mount the Grid.js to */
  container?: Element;
  /** gridjs-temp div which is used internally */
  tempRef?: RefObject<HTMLDivElement>;
  data?:
    | TwoDArray<TCell>
    | (() => TwoDArray<TCell>)
    | (() => Promise<TwoDArray<TCell>>);
  server?: ServerStorageOptions;
  header?: Header;
  /** to parse a HTML table and load the data */
  from: HTMLElement;
  storage: Storage<any>;
  pipeline: Pipeline<Tabular>;
  /** to automatically calculate the columns width */
  autoWidth: boolean;
  /** sets the width of the container and table */
  width: string;
  search: SearchConfig;
  pagination: PaginationConfig;
  sort: GenericSortConfig;
  translator: Translator;
}

// Config type used by the consumers
interface UserConfigExtend {
  columns?: OneDArray<TColumn | string>;
  search: SearchConfig | boolean;
  pagination: PaginationConfig | boolean;
  // implicit option to enable the sort plugin globally
  sort: GenericSortConfig | boolean;
  language: Language;
}

export type UserConfig = ProtoExtends<
  Partial<Config>,
  Partial<UserConfigExtend>
>;

export class Config {
  constructor(config?: Config) {
    // FIXME: not sure if this makes sense because Config is a subset of UserConfig
    const updatedConfig = {
      ...Config.defaultConfig(),
      ...config,
    };

    Object.assign(this, updatedConfig);
  }

  update(updatedConfig: Partial<Config>): Config {
    Object.assign(this, updatedConfig || {});
    return this;
  }

  static defaultConfig(): Config {
    return {
      tempRef: createRef(),
      width: '100%',
      autoWidth: true,
    } as Config;
  }

  static fromUserConfig(userConfig: UserConfig): Config {
    const config = new Config(userConfig as Config);

    config.update({
      dispatcher: new Dispatcher<any>(),
      storage: StorageUtils.createFromUserConfig(userConfig),
    });

    config.update({
      pipeline: PipelineUtils.createFromConfig(config),
    });

    // Sort
    if (typeof userConfig.sort === 'boolean' && userConfig.sort) {
      config.update({
        sort: {
          multiColumn: true,
        },
      });
    }

    // Header
    config.update({
      header: Header.fromUserConfig(config),
    });

    // Pagination
    config.update({
      pagination: {
        enabled:
          userConfig.pagination === true ||
          userConfig.pagination instanceof Object,
        ...(userConfig.pagination as PaginationConfig),
      },
    });

    // Search
    config.update({
      search: {
        enabled:
          userConfig.search === true || userConfig.search instanceof Object,
        ...(userConfig.search as SearchConfig),
      },
    });

    // Translator
    config.update({
      translator: new Translator(userConfig.language),
    });

    return config;
  }
}
