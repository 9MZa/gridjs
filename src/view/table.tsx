import { h } from 'preact';

import Tabular from '../tabular';
import { TBody } from './tbody';
import { THead } from './thead';
import { BaseComponent, BaseProps } from './base';
import Header from '../header';
import className from '../util/className';
import Config from '../config';
import { TBodyCell } from '../types';

import '../theme/mermaid/table.scss';

interface TableProps extends BaseProps {
  data?: Tabular<TBodyCell>;
  header?: Header;
}

export class Table extends BaseComponent<TableProps, {}> {
  render() {
    return (
      <table
        className={className(Config.current.classNamePrefix, 'table')}
        style={{ width: Config.current.width }}
      >
        <THead header={this.props.header} />
        <TBody data={this.props.data} />
      </table>
    );
  }
}
