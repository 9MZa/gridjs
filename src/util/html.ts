import { h, VNode } from 'preact';
import { HTMLContent } from '../view/htmlElement';

export function html(content: string, parentElement?: string): VNode {
  return h(HTMLContent, { content: content, parentElement: parentElement });
}
