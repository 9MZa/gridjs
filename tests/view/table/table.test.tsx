import { mount } from 'enzyme';
import { createContext, h } from 'preact';
import { Table } from '../../../src/view/table/table';
import Header from '../../../src/header';
import { Config } from '../../../src/config';
import StorageUtils from '../../../src/storage/storageUtils';
import Pipeline from '../../../src/pipeline/pipeline';
import StorageExtractor from '../../../src/pipeline/extractor/storage';
import ArrayToTabularTransformer from '../../../src/pipeline/transformer/arrayToTabular';
import { Status, TCell } from '../../../src/types';
import Dispatcher from '../../../src/util/dispatcher';
import { Translator } from '../../../src/i18n/language';
import Tabular from '../../../src/tabular';
import { html } from '../../../src/util/html';

describe('Table component', () => {
  let config: Config;
  const configContext = createContext(null);

  beforeEach(() => {
    config = new Config();
    config.data = [
      [1, 2, 3],
      ['a', 'b', 'c'],
    ];

    config.autoWidth = true;
    config.storage = StorageUtils.createFromUserConfig(config);
    config.dispatcher = new Dispatcher();
    config.translator = new Translator();
    config.pipeline = new Pipeline([
      new StorageExtractor({ storage: config.storage }),
      new ArrayToTabularTransformer(),
    ]);
  });

  it('should render a table', async () => {
    const table = mount(
      <configContext.Provider value={config}>
        <Table
          data={await config.pipeline.process()}
          status={Status.Loaded}
          width={config.width}
          height={config.height}
        />
      </configContext.Provider>,
    );

    expect(table.html()).toMatchSnapshot();
  });

  it('should render a table with loading', async () => {
    const table = mount(
      <configContext.Provider value={config}>
        <Table
          data={await config.pipeline.process()}
          status={Status.Loading}
          width={config.width}
          height={config.height}
        />
      </configContext.Provider>,
    );

    expect(table.html()).toMatchSnapshot();
  });

  it('should render a table with header', async () => {
    const table = mount(
      <configContext.Provider value={config}>
        <Table
          data={await config.pipeline.process()}
          header={Header.fromUserConfig({ columns: ['h1', 'h2', 'h3'] })}
          status={Status.Loaded}
          width={config.width}
          height={config.height}
        />
      </configContext.Provider>,
    );

    expect(table.html()).toMatchSnapshot();
  });

  it('should render a table with width', async () => {
    const table = mount(
      <configContext.Provider value={config}>
        <Table
          data={await config.pipeline.process()}
          width="300px"
          height={config.height}
          header={Header.fromUserConfig({ columns: ['h1', 'h2', 'h3'] })}
          status={Status.Loaded}
        />
      </configContext.Provider>,
    );

    expect(table.html()).toMatchSnapshot();
  });

  it('should render a table with column width', async () => {
    const header = Header.fromUserConfig({ columns: ['h1', 'h2', 'h3'] });
    header.columns[0].width = '10%';
    header.columns[2].width = '300px';

    const table = mount(
      <configContext.Provider value={config}>
        <Table
          data={await config.pipeline.process()}
          header={header}
          status={Status.Loaded}
          width={config.width}
          height={config.height}
        />
      </configContext.Provider>,
    );

    expect(table.html()).toMatchSnapshot();
  });

  it('should render a table with column sort', async () => {
    const header = Header.fromUserConfig({ columns: ['h1', 'h2', 'h3'] });
    header.columns[0].sort = {
      enabled: true,
    };
    header.columns[2].sort = {
      enabled: true,
    };

    const table = mount(
      <configContext.Provider value={config}>
        <Table
          data={await config.pipeline.process()}
          header={header}
          status={Status.Loaded}
          width={config.width}
          height={config.height}
        />
      </configContext.Provider>,
    );

    expect(table.html()).toMatchSnapshot();
  });

  it('should render a table without autoFix', async () => {
    const header = Header.fromUserConfig({
      columns: ['h1', 'h2', 'h3'],
      autoWidth: false,
    });

    const table = mount(
      <configContext.Provider value={config}>
        <Table
          data={await config.pipeline.process()}
          header={header}
          status={Status.Loaded}
          width={config.width}
          height={config.height}
        />
      </configContext.Provider>,
    );

    expect(table.html()).toMatchSnapshot();
  });

  it('should render a table without data', async () => {
    const header = Header.fromUserConfig({
      columns: ['h1', 'h2', 'h3'],
    });

    const table = mount(
      <configContext.Provider value={config}>
        <Table
          data={Tabular.fromArray<TCell>([])}
          header={header}
          status={Status.Loaded}
          width={config.width}
          height={config.height}
        />
      </configContext.Provider>,
    );

    expect(table.html()).toMatchSnapshot();
  });

  it('should render a table with null', async () => {
    const header = Header.fromUserConfig({
      columns: ['h1', 'h2', 'h3'],
    });

    const table = mount(
      <configContext.Provider value={config}>
        <Table
          data={null}
          header={header}
          status={Status.Loaded}
          width={config.width}
          height={config.height}
        />
      </configContext.Provider>,
    );

    expect(table.html()).toMatchSnapshot();
  });

  it('should attach styles', async () => {
    const header = Header.fromUserConfig({
      columns: ['h1', 'h2', 'h3'],
    });

    config.style = {
      th: {
        border: '1px solid black',
      },
      table: {
        padding: '2px',
        margin: '1px',
      },
    };

    const table = mount(
      <configContext.Provider value={config}>
        <Table
          data={null}
          header={header}
          status={Status.Loaded}
          width={config.width}
          height={config.height}
        />
      </configContext.Provider>,
    );

    expect(table.html()).toMatchSnapshot();
  });

  it('should render header with complex content', async () => {
    const header = Header.fromUserConfig({
      columns: [html('<h1>h1</h1>'), 'h2', html('<b>h3</b>')],
    });

    const table = mount(
      <configContext.Provider value={config}>
        <Table
          data={null}
          header={header}
          status={Status.Loaded}
          width={config.width}
          height={config.height}
        />
      </configContext.Provider>,
    );

    expect(table.html()).toMatchSnapshot();
  });

  it('should render header with TColumn object', async () => {
    const header = Header.fromUserConfig({
      columns: [
        {
          name: html('<h1>h1</h1>'),
          id: 'boo',
        },
        'h2',
        {
          name: 'h3',
        },
      ],
    });

    const table = mount(
      <configContext.Provider value={config}>
        <Table
          data={null}
          header={header}
          status={Status.Loaded}
          width={config.width}
          height={config.height}
        />
      </configContext.Provider>,
    );

    expect(table.html()).toMatchSnapshot();
  });
});
