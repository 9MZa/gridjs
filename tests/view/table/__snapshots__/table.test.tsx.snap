// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table component should render a table 1`] = `"<table class=\\"gridjs-table\\"><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">1</td><td class=\\"gridjs-td\\">2</td><td class=\\"gridjs-td\\">3</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">a</td><td class=\\"gridjs-td\\">b</td><td class=\\"gridjs-td\\">c</td></tr></tbody></table>"`;

exports[`Table component should render a table with column sort 1`] = `"<table class=\\"gridjs-table\\"><thead class=\\"gridjs-thead\\"><tr class=\\"gridjs-tr\\"><th class=\\"gridjs-th gridjs-th-sort\\">h1<button title=\\"Sort column ascending\\" class=\\"gridjs-sort gridjs-sort-neutral\\"></button></th><th class=\\"gridjs-th\\">h2</th><th class=\\"gridjs-th gridjs-th-sort\\">h3<button title=\\"Sort column ascending\\" class=\\"gridjs-sort gridjs-sort-neutral\\"></button></th></tr></thead><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">1</td><td class=\\"gridjs-td\\">2</td><td class=\\"gridjs-td\\">3</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">a</td><td class=\\"gridjs-td\\">b</td><td class=\\"gridjs-td\\">c</td></tr></tbody></table>"`;

exports[`Table component should render a table with column width 1`] = `"<table class=\\"gridjs-table\\"><thead class=\\"gridjs-thead\\"><tr class=\\"gridjs-tr\\"><th class=\\"gridjs-th\\" style=\\"width: 10%;\\">h1</th><th class=\\"gridjs-th\\">h2</th><th class=\\"gridjs-th\\" style=\\"width: 300px;\\">h3</th></tr></thead><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">1</td><td class=\\"gridjs-td\\">2</td><td class=\\"gridjs-td\\">3</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">a</td><td class=\\"gridjs-td\\">b</td><td class=\\"gridjs-td\\">c</td></tr></tbody></table>"`;

exports[`Table component should render a table with header 1`] = `"<table class=\\"gridjs-table\\"><thead class=\\"gridjs-thead\\"><tr class=\\"gridjs-tr\\"><th class=\\"gridjs-th\\">h1</th><th class=\\"gridjs-th\\">h2</th><th class=\\"gridjs-th\\">h3</th></tr></thead><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">1</td><td class=\\"gridjs-td\\">2</td><td class=\\"gridjs-td\\">3</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">a</td><td class=\\"gridjs-td\\">b</td><td class=\\"gridjs-td\\">c</td></tr></tbody></table>"`;

exports[`Table component should render a table with loading 1`] = `"<table class=\\"gridjs-table\\"><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">1</td><td class=\\"gridjs-td\\">2</td><td class=\\"gridjs-td\\">3</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">a</td><td class=\\"gridjs-td\\">b</td><td class=\\"gridjs-td\\">c</td></tr></tbody></table>"`;

exports[`Table component should render a table with width 1`] = `"<table class=\\"gridjs-table\\" style=\\"width: 300px;\\"><thead class=\\"gridjs-thead\\"><tr class=\\"gridjs-tr\\"><th class=\\"gridjs-th\\">h1</th><th class=\\"gridjs-th\\">h2</th><th class=\\"gridjs-th\\">h3</th></tr></thead><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">1</td><td class=\\"gridjs-td\\">2</td><td class=\\"gridjs-td\\">3</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">a</td><td class=\\"gridjs-td\\">b</td><td class=\\"gridjs-td\\">c</td></tr></tbody></table>"`;

exports[`Table component should render a table without autoFix 1`] = `"<table class=\\"gridjs-table\\"><thead class=\\"gridjs-thead\\"><tr class=\\"gridjs-tr\\"><th class=\\"gridjs-th\\">h1</th><th class=\\"gridjs-th\\">h2</th><th class=\\"gridjs-th\\">h3</th></tr></thead><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">1</td><td class=\\"gridjs-td\\">2</td><td class=\\"gridjs-td\\">3</td></tr><tr class=\\"gridjs-tr\\"><td class=\\"gridjs-td\\">a</td><td class=\\"gridjs-td\\">b</td><td class=\\"gridjs-td\\">c</td></tr></tbody></table>"`;

exports[`Table component should render a table without data 1`] = `"<table class=\\"gridjs-table\\"><thead class=\\"gridjs-thead\\"><tr class=\\"gridjs-tr\\"><th class=\\"gridjs-th\\">h1</th><th class=\\"gridjs-th\\">h2</th><th class=\\"gridjs-th\\">h3</th></tr></thead><tbody class=\\"gridjs-tbody\\"><tr class=\\"gridjs-tr\\"><td colspan=\\"3\\" class=\\"gridjs-td gridjs-message gridjs-notfound\\">No matching records found</td></tr></tbody></table>"`;
